mongoengine/base/document.py

```diff
@@ -478,7 +478,20 @@ def _mark_as_changed(self, key):
             key = self._db_field_map.get(key, key)
 
         if key not in self._changed_fields:
-            self._changed_fields.append(key)
+            levels = key.split('.')
+            idx = 1
+            while idx <= len(levels):
+                if '.'.join(levels[:idx]) in self._changed_fields:
+                    break
+                idx += 1
+            else:
+                self._changed_fields.append(key)
+                # remove lower level changed fields
+                level = '.'.join(levels) + '.'
+                remove = self._changed_fields.remove
+                for field in self._changed_fields[:]:
```

Comment:
yeah, i thought it wrong, i mix it up with the dict that cannot change the instance when iterating